src/main/java/org/commcare/formplayer/beans/menus/QueryResponseBean.java

```diff
@@ -43,9 +44,17 @@ public QueryResponseBean(QueryScreen queryScreen, SessionWrapper session) {
             String currentAnswer = currentAnswers.get(key);
 
             // Map the current Answer to the itemset index of the answer
-            Pair<String[], Integer> choicesAndAnswerIndex = queryPromptItem.getItemsetChoicesWithAnswerIndex(currentAnswer);
-            if (queryPromptItem.isSelectOne()) {
-                currentAnswer = choicesAndAnswerIndex.second == -1 ? null : String.valueOf(choicesAndAnswerIndex.second);
+            String[] choiceLabels = null;
+            if (queryPromptItem.isSelect()) {
+                String[] selectedChoices = RemoteQuerySessionManager.extractSelectChoices(currentAnswer);
+                String answerWithChoiceIndices = "";
+                for (int i = 0; i < selectedChoices.length; i++) {
+                    int choiceIndex = ItemSetUtils.getIndexOf(queryPromptItem.getItemsetBinding(), selectedChoices[i]);
+                    answerWithChoiceIndices += choiceIndex == -1 ? "" : String.valueOf(choiceIndex);
+                    answerWithChoiceIndices += i == selectedChoices.length -1 ? "" : RemoteQuerySessionManager.MULTI_SELECT_DELIMITER;
+                }
+                currentAnswer = answerWithChoiceIndices;
```

Comment:
quite a lot of duplication with `QueryScreen.answerPrompts`, any way to reduce that?

Reply:
I looked at it, but both of these places seem to be doing fundamentally opposite things making it hard to DRY this code out. The variable declarations and loop is same in both of these places though the value conversion is opposite and I don't have good ideas on how to unify this code better.